### PR TITLE
fix: Plan entity decoder not to be strict on types

### LINKED ISSUE #1259: Add support for schema V10
CLI is introducing a set of changes to the current schema and decided to release a new V10. KG has to catch up with them and do relevant changes in various parts of the services.

**Acceptance criteria:**

- [ ] immediate changes
  - [x] modify the re-provisioning kick-off conditions so the process won't be triggered for V10 (PR)
    - [x] replace the `compatibility-matrix` in the `application.conf` to `cli-version`, `schema-version`, `re-provisioning-needed=yes|no`
    - [x] if cliVersion and schemaVersion of CLI and TS are the same, don't reprovision (even if `re-provisioning-needed=yes`)
    - [x] if cliVersion (or schemaVersion) of CLI and TS are different AND `re-provisioning-needed=yes`, kick-off the process
    - [x] if cliVersion (or schemaVersion) of CLI and TS are different AND `re-provisioning-needed=no`, do nothing
    - [x] fail TG to start if either cliVersion or schemaVersion defined in the `application.conf` and read from the CLI are different (assuming CLI can tell us what schemaVersion it produces)
  - [x] deal with DS `dateCreated` and `dateModified` (no re-provisioning needed)
  ```
  DS type         init                modification (new)                  modification (old)            KG approach
  internal DS ->  dateCreated      -> dateCreated + dateModified       -> dateCreated                   map dateModified to dateCreated
  internal DS ->  dateCreated      -> dateCreated + invalidationTime   -> dateCreated, invalidationTime map invalidationTime to dateCreated, invalidationTime
  external DS ->  datePublished    -> datePublished + dateModified     -> dateCreated                   map dateModified to dateCreated
  external DS ->  datePublished    -> datePublished + invalidationTime -> dateCreated, invalidationTime map invalidationTime to dateCreated, invalidationTime
  ```
  - [x] deal with Plan `dateCreated` and `dateModified` (no re-provisioning needed)
    - [x] for Plan modifications just map `dateModified` to `dateCreated`
  - [x] make sure strict type matching is used for `Plan` and `CompositePlan` decoders
  - [x] make sure the `ParameterMapping`'s `defaultValue` is optional
  - [x] do we want to wait for the shacl validation change on the defaultValue
  - [x] make sure CLI encodes URL objects still;
  - [x] double check if we have this new slug-like name on the Plan entity; if yes, do relevant changes; double check how this will work for the old Plans during `renku migrate`; hopefully, the new property value will be a copy of the old `name`; if not, we'd need to re-provision all the projects having Plans
  - [x] double check if `invalidationTime` format on either Dataset or Plan is causing us any decoding problems

### PR DESCRIPTION
**Pull request by jachro**
This PR takes out the strict type check predicate from both the Plan entity decoders. The reason is that introduction of `WorkflowFile` version of both the Plans does not bring significant changes to any of the non-`WorkflowFile` Plans. Moreover, not decoding `WorkflowFile` Plans causes failures during decoding Activities created as a result of `WorkflowFile` Plan execution. The option here would be to skip decoding `WorkflowFile` related Activities but this seems not to be a trivial thing to do.
I tried decoding a payload from a project using WorkflowFile feature and without the strict type check on Plan decoders all seems to be working fine.

### CODE REVIEW

File: renku-model/src/test/scala/io/renku/graph/model/entities/PlanSpec.scala [hunk 1]
```diff
@@ -186,45 +187,15 @@ class PlanSpec
...
-      val newJson =
-        JsonLDTools
-          .view(sp)
-          .selectByTypes(entities.StepPlan.entityTypes)
-          .addType(renku / "WorkflowPlan")
-          .value
+      val decoded = jsonLD.cursor.as[List[entities.Plan]]
 
-      newJson.cursor.as[List[entities.StepPlan]] shouldBe Right(Nil)
+      decoded.fold(err => fail(err.message), identity) should contain theSameElementsAs List(sp, cp)
```
Comment (eikek) [inline renku-model/src/test/scala/io/renku/graph/model/entities/PlanSpec.scala:198]: A test for the case we have an additional type in the jsonld would also be useful, no? like the removed `fail decode if…` tests (that didn't really fail, but returned an empty list) but on the `Plan.decoder` level (not on the subtypes) - iiuc that was the problem yesterday?

File: renku-model/src/main/scala/io/renku/graph/model/entities/Plan.scala [hunk 2]
```diff
@@ -61,21 +61,15 @@ object Plan {
...
-        }
+    JsonLDDecoder.entity(EntityTypes of prov / "Plan") { cursor =>
+      lazy val noMatchFailure: Either[DecodingFailure, Plan] = cursor.getEntityTypes.flatMap { ets =>
+        DecodingFailure(
+          DecodingFailure.Reason.CustomReason(show"Cannot decode entity as Plan: $ets"),
+          cursor.jsonLD.toJson.hcursor
+        ).asLeft
+      }
+
+      StepPlan.decoder.apply(cursor) orElse CompositePlan.decoder.apply(cursor) orElse noMatchFailure
```
Comment (eikek) [inline renku-model/src/main/scala/io/renku/graph/model/entities/Plan.scala:72]: Interesting! I _think_ I tried a similar solution at first but it failed some tests I created 🤔  (maybe the tests were not good 😄)

Comment (jachro) [reply renku-model/src/main/scala/io/renku/graph/model/entities/Plan.scala:72]: Hm, don't really know what to say

File: renku-model/src/test/scala/io/renku/graph/model/entities/PlanSpec.scala (hunk 1 above)
Comment (jachro) [reply renku-model/src/test/scala/io/renku/graph/model/entities/PlanSpec.scala:198]: Yes, we could have one. Let me think if I can compile something like that.

File: renku-model/src/main/scala/io/renku/graph/model/entities/Plan.scala (hunk 2 above)
Comment (eikek) [reply renku-model/src/main/scala/io/renku/graph/model/entities/Plan.scala:72]: Sorry, it was just a thought in my head, nothing to answer :)

**Review by eikek — APPROVED** [review]
